src/host/readDataCooked.cpp

```diff
@@ -436,633 +196,1027 @@ bool COOKED_READ_DATA::AtEol() const noexcept
+void COOKED_READ_DATA::_erase(const til::CoordType distance)
+{
+    if (distance > 0)
+    {
+        const std::wstring str(gsl::narrow_cast<size_t>(distance), L' ');
+        _writeChars(str);
+    }
+}
+
+// A helper to write text while keeping track of buffer scrolling. When writing text in the last line of the
+// buffer, all of the text will shift upwards, which means that our cached coordinates need to shift as well.
+void COOKED_READ_DATA::_writeChars(const std::wstring_view& text)
+{
+    if (text.empty())
     {
-        size_t NumSpaces = 0;
-        til::CoordType ScrollY = 0;
-
-        FAIL_FAST_IF_NTSTATUS_FAILED(WriteCharsLegacy(ScreenInfo(),
-                                                      _backupLimit,
-                                                      _bufPtr,
-                                                      _bufPtr,
-                                                      &bytesInserted,
-                                                      &NumSpaces,
-                                                      OriginalCursorPosition().x,
-                                                      WC_INTERACTIVE | WC_KEEP_CURSOR_VISIBLE,
-                                                      &ScrollY));
-        OriginalCursorPosition().y += ScrollY;
-        VisibleCharCount() += NumSpaces;
+        return;
     }
-    _bufPtr += charsInserted;
 
-    return charsInserted;
+    til::CoordType scrollY = 0;
+    WriteCharsLegacy(_screenInfo, text, true, &scrollY);
+    _coordBeg.y = std::max(0, _coordBeg.y + scrollY);
+    _coordCursor.y = std::max(0, _coordCursor.y + scrollY);
+    _coordEnd.y = std::max(0, _coordEnd.y + scrollY);
 }
 
-// Routine Description:
-// - saves data in the prompt buffer to the outgoing user buffer
-// Arguments:
-// - cch - the number of chars to write to the user buffer
-// Return Value:
-// - the number of bytes written to the user buffer
-size_t COOKED_READ_DATA::SavePromptToUserBuffer(const size_t cch)
+// Just a simple helper to replace the entire buffer contents.
+void COOKED_READ_DATA::_replaceBuffer(const std::wstring_view& str)
+{
+    _buffer.assign(str);
+    _bufferCursor = _buffer.size();
+    _markAsDirty();
+}
+
+// If the viewport is large enough to fit a popup, this function prepares everything for
+// showing the given type. It handles computing the size of the popup, its position,
+// backs the affected area up and draws the border and initial contents.
+void COOKED_READ_DATA::_popupPush(const PopupKind kind)
+try
 {
-    size_t bytesToWrite = 0;
-    const auto hr = SizeTMult(cch, sizeof(wchar_t), &bytesToWrite);
-    if (FAILED(hr))
+    auto& textBuffer = _screenInfo.GetTextBuffer();
+    const auto viewport = _screenInfo.GetViewport();
+    const auto viewportOrigin = viewport.Origin();
+    const auto viewportSize = viewport.Dimensions();
+
+    til::size proposedSize;
+    switch (kind)
+    {
+    case PopupKind::CopyToChar:
+        proposedSize = { 26, 1 };

```

Comment:
okay so this used to be 26+2  (in `CopyToCharPopup.cpp`). Most of them were all `{{number+2}}``. Except the `CommandNumber` one - that was 22+5. 

Why are these now okay being just `{{number}}` (and not `number+2`)

Reply:
The old code used the outer dimensions (i.e. with the frame) to construct a `Prompt`. The new code uses the inner dimensions. I felt like that this makes the code simpler because now the popup-specific code only needs to compute however much space it needs and the popup-unspecific code calculates the outer dimensions automatically. It also helps constrain the text drawing inside the popup since it needs to be clipped to the inner dimensions.